### PR TITLE
fix: token_spend calculation incorrect with concurrent webhooks

### LINKED ISSUE #970: bug: token_spend calculation incorrect when concurrent webhooks share the same GitHub token
## Description

The `token_spend` calculation is incorrect when multiple webhooks run concurrently using the same GitHub API token.

## Root Cause

The `CountingRequester` wrapper is **shared** across webhook instances when they reuse the same GitHub API client:

```python
# In GithubWebhook.__init__()
if isinstance(requester, CountingRequester):
    # Reuse existing wrapper (SHARED across webhooks!)
    self.requester_wrapper = requester
```

## Impact

When 2+ webhooks run concurrently with the same token:
1. Both record `initial_count = 100` at start
2. Webhook A makes 5 calls, Webhook B makes 3 calls → shared counter = 108
3. Both calculate `token_spend = 108 - 100 = 8` (incorrect for both)

The `initial_rate_limit` and `final_rate_limit` fields are also affected.

## Location

- `webhook_server/libs/github_api.py` - `GithubWebhook.__init__()` and `_update_context_metrics()`

## Possible Fixes

1. **Per-webhook tracking** - Create a new CountingRequester per webhook instead of sharing
2. **Atomic snapshot** - Track per-webhook deltas using context-based counters
3. **Rate limit delta** - Use actual GitHub rate limit API at start/end (adds 2 API calls overhead)

## Labels

- bug
- enhancement

### PR DESCRIPTION
**Pull request by myakove**
## Problem

`CountingRequester` was broken in two ways:

1. **`withLazy()` bypass**: PyGithub's `Github.get_repo()` calls `requester.withLazy()` which returns a NEW raw `Requester`, completely bypassing the `CountingRequester` wrapper. Only `Github`-level calls (`get_rate_limit`) were counted — all Repository-level API calls (labels, check runs, comments) were invisible. A PR that ran for 2m50s doing dozens of operations reported only **1 API call**.

2. **Shared counter between concurrent webhooks**: If a `Github` instance were reused, multiple webhooks would share the same `CountingRequester` and corrupt each other's `token_spend`.

## Fix

- **Override `withLazy()`** in `CountingRequester` to return a new wrapper sharing the same counter (`_shared_count` list + `_thread_lock`)
- **Always create a fresh `CountingRequester` per webhook** — unwrap any existing wrapper to get the real requester
- **Remove `initial_wrapper_count`** — each webhook starts at 0, no delta needed

## Also includes

- **Single-token optimization**: Skip rate limit comparison loop when only 1 token is configured (saves 1 API call per webhook)
- **`WEBHOOK_SERVER_DEV_MODE`** env var for uvicorn auto-reload during development
- **Remove `hasattr(_Github__requester)`** guard — PyGithub >=2.4.0 guaranteed, direct access per AGENTS.md fail-fast principle

## Verified on live dev server

| Metric | Before fix | After fix |
|--------|-----------|-----------|
| pull_request API calls | 1 ❌ | 31 ✅ |
| Rate limit math | 3934-3933=1 | 3340-3309=31 ✅ |
| Rate limit capacity logged | ✅ | ✅ |

## Testing

- All 39 related tests pass
- ruff check/format clean
- Peer-reviewed by Cursor (2 rounds)

Closes #970

## Summary by CodeRabbit

* **New Features**
  * Dev-mode server reload via environment toggle; worker/process behavior adjusted at runtime.
  * Shared request-counting wrappers for consistent API metrics.

* **Bug Fixes**
  * Improved single-token API validation and short-circuit handling.
  * More accurate token/rate-limit accounting using absolute request counts; prevents cross-instance counter sharing.

* **Tests**
  * Expanded coverage for metrics, single-token flows, init/destructor edge cases and concurrent webhooks.

* **Documentation**
  * Updated token-spend description in webhook context.

### CODE REVIEW

**Comment by myakove** [comment]
@coderabbitai review

File: entrypoint.py [hunk 1]
```diff
@@ -58,5 +60,6 @@ def run_podman_cleanup() -> None:
         host=_ip_bind,
         port=int(_port),
         workers=int(_max_workers),
-        reload=False,
+        # Note: reload=True forces single-worker mode (uvicorn limitation)
+        reload=_dev_mode,
```
[the remark this answers is not in the document]
Comment (myakove) [reply entrypoint.py]: Fixed: workers parameter is now only passed when reload is disabled (dev mode off).

File: webhook_server/libs/github_api.py [hunk 2]
```diff
@@ -195,14 +214,14 @@ async def _update_context_metrics(self) -> None:
...
 
         if self.initial_rate_limit_remaining is not None:
             self.ctx.initial_rate_limit = self.initial_rate_limit_remaining
             if self.requester_wrapper:
-                self.ctx.final_rate_limit = max(
-                    0, self.initial_rate_limit_remaining - (self.requester_wrapper.count - self.initial_wrapper_count)
-                )
+                # Heuristic: assumes each REST call consumes one rate limit unit.
+                # Does not account for GitHub App client calls or endpoint-specific costs.
+                self.ctx.final_rate_limit = max(0, self.initial_rate_limit_remaining - self.requester_wrapper.count)
```
[the remark this answers is not in the document]
Comment (myakove) [reply webhook_server/libs/github_api.py:224]: Fixed: Updated WebhookContext.token_spend docstring to reflect absolute count from requester wrapper.

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

File: webhook_server/tests/test_github_api_metrics.py [hunk 3]
```diff
@@ -280,8 +308,13 @@ async def test_get_token_metrics_fallback_reset(
...
+
+        __del__ uses getattr with a default because it can be called during failed
+        initialization when clone_repo_dir was never set.
+        """
         gh = GithubWebhook.__new__(GithubWebhook)
-        # Should not raise AttributeError
-        gh.__del__()
+        # Simulate partially initialized instance (no clone_repo_dir set)
+        # __del__ should handle this via getattr(..., None) without raising
+        del gh
```
[the remark this answers is not in the document]
Comment (myakove) [reply webhook_server/tests/test_github_api_metrics.py]: Fixed: Changed to GithubWebhook.__del__(gh) for deterministic exception surfacing.

File: webhook_server/utils/helpers.py [hunk 4]
```diff
@@ -492,6 +492,23 @@ def get_api_with_highest_rate_limit(config: Config, repository_name: str = "") -
...
+        try:
+            _api_user = _api.get_user().login
+        except GithubException as ex:
+            raise NoApiTokenError(f"Single configured token is invalid: {ex}") from ex
+
+        _rate_limit = _api.get_rate_limit()
+        log_rate_limit(rate_limit=_rate_limit, api_user=_api_user)
+
+        logger.info(f"API user {_api_user} selected (single token configured)")
+        return _api, _token, _api_user
```
[the remark this answers is not in the document]
Comment (myakove) [reply webhook_server/utils/helpers.py:510]: Skipped: The user explicitly requested rate limit capacity (xxx/5000) to remain in the logs. Removing get_rate_limit() would remove those logs.

**Comment by myakove** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `webhook_server/tests/test_github_api_metrics.py:145` (outside diff) — HIGH: This still doesn't lock in the issue `#970` regression.
> Addressed: Fixed: Added test_concurrent_webhooks_have_independent_counts regression test proving two GithubWebhook instances have independent counts.

**Comment by myakove** [comment]
@coderabbitai review

File: entrypoint.py [hunk 5]
```diff
@@ -14,6 +16,7 @@
 _port = _root_config.get("port", 5000)
 _max_workers = _root_config.get("max-workers", 10)
 _webhook_secret = _root_config.get("webhook-secret")
+_dev_mode = os.environ.get("WEBHOOK_SERVER_DEV_MODE", "").lower() in ("1", "true", "yes")
```
[the remark this answers is not in the document]
Comment (myakove) [reply entrypoint.py:19]: Fixed: Added WEBHOOK_SERVER_DEV_MODE to the runtime env vars list in docs/docker-deployment.md.

File: webhook_server/tests/test_helpers.py [hunk 6]
```diff
@@ -156,6 +156,45 @@ def test_get_api_with_highest_rate_limit_invalid_tokens(
...
+
+        mock_get_apis.return_value = [(mock_api, "token1")]
+
+        with patch.dict(os.environ, {"WEBHOOK_SERVER_DATA_DIR": "webhook_server/tests/manifests"}):
+            config = Config(repository="test-repo")
+            api, token, user = get_api_with_highest_rate_limit(config=config, repository_name="test-repo")
+
+        assert api == mock_api
+        assert token == "token1"
+        assert user == "user1"
```
[the remark this answers is not in the document]
Comment (myakove) [reply webhook_server/tests/test_helpers.py:180]: Skipped: S105 is not enabled in ruff config and detect-secrets does not flag these short test strings (pre-commit hooks pass). Pre-existing test tokens at lines 80/148 use the same pattern without pragmas.

**Comment by myakove** [comment]
@coderabbitai review
